packages/datetime/src/datePicker.tsx

```diff
@@ -162,22 +129,20 @@ export class DatePicker extends AbstractPureComponent<IDatePickerProps, IDatePic
     }
 
     public componentWillReceiveProps(nextProps: IDatePickerProps) {
-        if (nextProps.value !== this.props.value) {
-            let { displayMonth, displayYear, selectedDay } = this.state;
-            if (nextProps.value != null) {
-                displayMonth = nextProps.value.getMonth();
-                displayYear = nextProps.value.getFullYear();
-                selectedDay = nextProps.value.getDate();
+        super.componentWillReceiveProps(nextProps);
+        const { value } = nextProps;
+        if (value !== this.props.value) {
```

Comment:
nit: to avoid deep nesting in these situations, I prefer to handle easy/erroneous cases first to achieve a flatter, easier-to-read code structure:

```tsx
if (value === this.props.value) {
    // no action needed
    return;
}

if (value == null) {
    // clear the value
    this.setState({ value });
}

// otherwise, use the new value
this.setState({
    displayMonth: value.getMonth(),
    displayYear: value.getFullYear(),
    selectedDay: value.getDate(),
    value,
});
```